Resource picker/improve ux

nmarrs opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
Replace modal based UX with popover UX similar to color picker. This makes UX of selection more uniform across the board (between selecting colors / icons) and is a first step in including resource picker in context of value mapping.

Note: planning on writing tests for this as part of including in context of value mapping (next PR)

Before
https://user-images.githubusercontent.com/22381771/150873092-6938171e-c50f-4350-bf61-f82d7d58bc56.mov

After
Initial UX
https://user-images.githubusercontent.com/22381771/150872451-e9c09b2f-a797-4205-be32-955486750629.mov

NEW updated UX

https://user-images.githubusercontent.com/22381771/151071786-36a43e99-f76d-4eb2-8f53-cbd67b03b3a0.mov




**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #40723

nmarrs committed:
First attempt at switching resource picker to using popover UX instead of modal

nmarrs committed:
Clean up resourcepicker popover component

nmarrs committed:
Cleanup resource picker component

nmarrs committed:
Fix broken import

radiorental commented:
Looks really good.  Does it need a [cancel] or [X] to close the dialog?

atifali reviewed (approved):
tested this on my dashboard - workflow is great - agree with @radiorental that some kind of exit method would be nice 👍

radiorental reviewed (approved):
ship it

radiorental commented:
@atifali I discussed with @nmarrs and decided to proceed without any form of cancel action on the panel.  Clicking elsewhere will close, no action from user will time-out the dialog.  It follows the current colour picker interaction paradigm.  We should keep an eye on it once it's out in the wild.  If users are struggling a little then adding a button in later should be relatively easy

dprokop commented on public/app/features/dimensions/editors/ResourcePickerPopover.tsx:
```diff
@@ -0,0 +1,273 @@
+      background: ${theme.colors.background.primary};
+      border: 1px solid ${theme.colors.border.medium};
+
+      .ResourcePickerPopover__tab {
+        width: 50%;
+        text-align: center;
+        padding: ${theme.spacing(1, 0)};
+        background: ${theme.colors.background.secondary};
+        color: ${theme.colors.text.secondary};
+        font-size: ${theme.typography.bodySmall.fontSize};
+        cursor: pointer;
+        border: none;
+
+        &:focus:not(:focus-visible) {
+          outline: none;
+          box-shadow: none;
+        }
+
+        :focus-visible {
+          position: relative;
+        }
+      }
+
+      .ResourcePickerPopover__tab--active {
+        color: ${theme.colors.text.primary};
+        font-weight: ${theme.typography.fontWeightMedium};
+        background: ${theme.colors.background.primary};
+      }
```
You don't need nesting in emotion, I would suggest going with a flat classes definition unless nesting serves here any purpose.

dprokop commented on public/app/features/dimensions/editors/ResourcePicker.tsx:
```diff
@@ -1,224 +1,93 @@
   );
 };
 
 const getStyles = stylesFactory((theme: GrafanaTheme2) => {
```
Please refactor stylesFactory to useStyles2. stylesFactory is should be used only when styles getter depend on some properties other than theme.

dprokop commented on public/app/features/dimensions/editors/ResourcePickerPopover.tsx:
```diff
@@ -0,0 +1,273 @@
+    }
+  };
+
+  const [newValue, setNewValue] = useState<string>(value ?? '');
```
A good practice with hooks is to defined them at the top level of the component. Otherwise you're in a reisk of introducing i.e. early exit before the hook.

nmarrs replied:
Sounds good. In this case I was trying to group similar logic which is one of the benefits of hooks over classes. In general does eng at Grafana prefer grouping all hooks at top of every file vs. grouping similar logic?

dprokop commented on public/app/features/dimensions/editors/ResourcePickerPopover.tsx:
```diff
@@ -0,0 +1,273 @@
+
+  const renderFolderPicker = () => (
+    <>
+      <Field>
+        <Select menuShouldPortal={true} options={folders} onChange={setCurrentFolder} value={currentFolder} />
+      </Field>
+      <Field>
+        <FilterInput
+          value={searchQuery ?? ''}
+          placeholder="Search"
+          onChange={(v) => {
+            onChangeSearch(v);
+            setSearchQuery(v);
+          }}
+        />
+      </Field>
+      {filteredIndex && (
+        <div className={styles.cardsWrapper}>
+          <ResourceCards cards={filteredIndex} onChange={(v) => setNewValue(v)} value={newValue} />
+        </div>
+      )}
+    </>
+  );
+
+  const renderURLPicker = () => (
+    <>
+      <Field>
+        <Input onChange={(e) => setNewValue(e.currentTarget.value)} value={newValue} />
+      </Field>
+      <div className={styles.iconContainer}>
+        <Field label="Preview">
+          <div className={styles.iconPreview}>
+            {mediaType === 'icon' && <SVG src={imgSrc} className={styles.img} />}
+            {mediaType === 'image' && newValue && <img src={imgSrc} className={styles.img} />}
+          </div>
+        </Field>
+        <Label>{shortName}</Label>
+      </div>
+    </>
+  );
```
Maybe these should be a separate components for the sake of this component complexity?

dprokop commented on public/app/features/dimensions/editors/ResourcePicker.tsx:
```diff
@@ -1,224 +1,93 @@
 
 interface Props {
   value?: string; //img/icons/unicons/0-plus.svg
+  srcPath?: string;
```
```suggestion
  src?: string;
```

dprokop commented on public/app/features/dimensions/editors/ResourcePicker.tsx:
```diff
@@ -1,224 +1,93 @@
 interface Props {
   value?: string; //img/icons/unicons/0-plus.svg
+  srcPath?: string;
+  niceName?: string;
```
```suggestion
  name?: string;
```

dprokop commented on public/app/features/dimensions/editors/ResourcePickerPopover.tsx:
```diff
@@ -0,0 +1,273 @@
+interface Props {
+  value?: string; //img/icons/unicons/0-plus.svg
+  onChange: (value?: string) => void;
+  mediaType: 'icon' | 'image';
```
Can you turn this typoing into an enum? I see it's being used in a few places in a similar way.

dprokop reviewed (commented):
Love the change! Left few comments to consider :)

The major doubt I have is the popover interaction. Since it's quite a complex interaction within the picker, I would expect it, as in the conversation you folks had, to have an explicit close button and not close on mouse out. Even as simple interaction as the one on the video below[1] leads to a stubbornly closing popover.

Maybe you could just disable the auto-close on mouse out and have an interaction similar to the one  we have with the dashboard time range picker - it closes either on user selection or outside click [2]

[1]

https://user-images.githubusercontent.com/2376619/150947861-87453a4d-2ec7-4c3d-938d-7bff191752a1.mov

[2]

https://user-images.githubusercontent.com/2376619/150948336-bddfe39c-3b82-4fd5-8779-3d0ae373b7ad.mov

nmarrs committed:
Update IconPicker UX so that it doesn't automatically close on mouse out, add cancel button, add click outside to close functionality

nmarrs committed:
Unnest popover tab css classes

nmarrs committed:
Refactor placement of hooks in ResoucePickerPopover file

nmarrs committed:
Refactor ResourcePicker prop naming

nmarrs committed:
Refactor mediaType into an enum

nmarrs committed:
Refactor styling to not use deprecated stylefactory module

nmarrs committed:
Refactor out folder / url tabs into separate components from resource picker popover

nmarrs commented:
Updated PR addressing all PR feedback left by @dprokop (thanks for the great feedback!)

Have updated the UX to match that of dashboard time range picker (i.e. no more mouse out timeout, rather popup can be closed by clicking outside the popup, selecting the newly added "cancel" button, or through selecting a new icon)

Would be great if someone could test this locally to make sure my recent refactors didn't have any unintended side affects that I may of missed. Thanks!

ryantxu commented on public/app/features/dimensions/editors/ResourcePicker.tsx:
```diff
@@ -1,224 +1,81 @@
-import { getDatasourceSrv } from 'app/features/plugins/datasource_srv';
-import { FileElement, GrafanaDatasource } from 'app/plugins/datasource/grafana/datasource';
-import { ResourceFolderName } from '..';
+import { MediaType, ResourceFolderName } from '..';
```
```suggestion
import { MediaType, ResourceFolderName } from '../types';
```
nit

atifali commented:
tested this again - nothing breaks on my end - nice work with the exit interaction ❤️

nmarrs committed:
Update type imports to be explicit
